Use data objects that are documented for the transaction object.

m50 opened the pull request:
When writing hooks, I noticed it was hard to figure out what I could modify and what I had access to, without digging through the Dredd docs.

While you can't get everything (you don't know what the body looks like, or what headers exist, for example), it helps a lot to have some static analyzability and documentation in editor.

This doesn't change any behaviour in the hooks, it just allows hooks to be read as documented objects.

This will mean, if you do `$transaction->` in PHPStorm or VSCode, you can see what values (and types) exist on the transaction object from in your editor.

This should be backwards compatible as well, since, the internal data structure hasn't changed.

All values are public properties, so that json_serialize can still read them, and so that they are accessed the same way.

Now, when writing hooks, if you write:

```php
<?php

use Dredd\Hooks;
use Dredd\DataObjects\Transaction;

Hooks::afterEach(function(Transaction &$transaction) {

});
```

You will now have full readability.

m50 committed:
Use data objects that are documented for the transaction.
This allows for in-IDE documentation and static analysis of the transaction objects on the hooks.

ddelnano commented:
@m50 sorry for getting back to you late on this and I really appreciate your contribution! I haven't had a chance to look into why the build is failing. I won't have time to dig into that until Friday but I will be reviewing this soon.

Making the experience of writing hooks easier through enhanced static analyzability is awesome, so I'm open to accepting this change. I just want to spend some time fixing the build before I do a more thorough review.

ddelnano commented:
@m50 please rebase on master now that I've fixed the majority of the issues with the test suite. The build still won't pass but I'll be able to do any additional verification now that master is in a much better state. More details on #46. 

Once your change is rebased I'll be able to quickly review this.

m50 commented:
@ddelnano sorry for the delay, been taking a break from the computer on the first week of my vacation, so just got to it now.

I rebased and re-pushed. Cucumber checks are still failing, as you said, but it's now matching master.

ddelnano commented on src/DataObjects/Request.php:
```diff
@@ -0,0 +1,44 @@
+
+    /**
+     * Keys are HTTP header names, values are HTTP header contents
+     * @var object $headers
```
This is an array of objects so I think this should be `@var array $headers`

m50 replied:
According to the docs, it's an object. Also when working with it, it was an object as well.

https://dredd.org/en/latest/data-structures.html#transaction-object

Take a look at request, it's an object there. Then again, this is for Javascript, so that may be where the discrepancy lies. 

It may be that it doesn't matter, and it can be an array, and if thats the case, I think it would be infinitely easier to work with keyed arrays than objects.

```php
$header = "Content-Type";
$transaction->request->$header = "application/json";
```

That's something I have had to do recently, and it's an awful way of working with headers, so hopefully you are right. I'll give it a test.

Tbf, I think when it's json encoded it'll be the same, the problem is when it's json decoded... Though I suppose in this case, we can cast in the end, so shouldn't matter. 

The description was copied from the Dredd docs.

ddelnano commented on src/DataObjects/Request.php:
```diff
@@ -0,0 +1,44 @@
+     * @var string $bodyEncoding
+     * @psalm-var 'utf-8'|'base64' $bodyEncoding
+     */
+    public $bodyEncoding;
```
I don't see where this is defined in the dredd docs [here](https://github.com/apiaryio/dredd-transactions#transaction-object) and I also don't see it on https://dredd.org/en/latest/hooks/index.html#hooks. Can you point out where this is documented? It seems to me the Dredd docs specify all the properties except `bodyEncoding`.

m50 replied:
It's in the data structures page.

https://dredd.org/en/latest/data-structures.html#transaction-object

Under "request" there it is.

ddelnano reviewed (changes requested):
Few minor comments. However, I would like to have a Cucumber test that uses the new function signature. You can find an example [here](https://github.com/ddelnano/dredd-hooks-php/blob/master/features/hook_handlers.feature). Let me know if you have any questions about implementing that. Running a single Cucumber test case with `npx cucumber-js features/name-of-your-feature-file.feature --require features/support/ --tags 'not @skip'`` from inside the project's Docker container would be the easiest way to run the test interactively.

m50 commented:
Ok, I'm out of town atm, so I'll make some cucumber tests when I get back. 👍

m50 committed:
Change headers to an array.
This allows for easier ability to work with it.
Cast the headers to an array in the construct, and define the type hint as a string=>string array.

m50 committed:
Install dredd into the docker container.
This is necessary, or else cucumber tests will fail.

m50 committed:
Add Transaction objects to tests, to validate it doesn't negatively affect anything.

m50 committed:
Should be `node` not `nodejs`.

m50 committed:
Add a cucumber test for Transaction objects.

m50 committed:
Undo Transaction objects, as these stdClass ones aren't well formed.

m50 committed:
Convert headers to an array of string=>string.

m50 committed:
Add some defaults, just in case.

m50 committed:
Add a test specifically for Transaction object.

m50 commented:
@ddelnano I added a phpunit test as well as a cucumber test.

Oddly, the cucumber test fails in the docker container, but not locally or in Travis. It complains about port already in use. Unsure what could be causing that, but figured I'd let you know.

Additionally, the docker container didn't have dredd installed in it, so to get the tests to run, I had to install dredd inside of it.

smujaddid commented:
Any update on this?

m50 commented:
> Any update on this?

Mostly just waiting for a response from @ddelnano. 

Dunno why it says that all checks have failed, considering last time I looked at this back in June, it was passing.
